diff --git a/TMessagesProj/build.gradle b/TMessagesProj/build.gradle
index ea8911bb1..8884b37fa 100644
--- a/TMessagesProj/build.gradle
+++ b/TMessagesProj/build.gradle
@@ -116,6 +116,13 @@ android {
     }
 
     buildTypes {
+        def localProperties = new Properties()
+        def localPropertiesFile = rootProject.file("local.properties")
+
+        if (localPropertiesFile.exists()) {
+            localProperties.load(new FileInputStream(localPropertiesFile))
+        }
+
         debug {
             jniDebuggable true
             minifyEnabled false
@@ -125,11 +132,13 @@ android {
             ndk.debugSymbolLevel = 'FULL'
             buildConfigField "String", "BUILD_VERSION_STRING", "\"" + APP_VERSION_NAME + "\""
             buildConfigField "String", "APP_CENTER_HASH", "\"\""
-            buildConfigField "String", "BETA_URL", "\"" + getProps("BETA_PRIVATE_URL") + "\""
+            buildConfigField "String", "BETA_URL", "\"" + localProperties.get("BETA_PRIVATE_URL", "") + "\""
             buildConfigField "boolean", "DEBUG_VERSION", "true"
             buildConfigField "boolean", "DEBUG_PRIVATE_VERSION", "true"
             buildConfigField "boolean", "BUILD_HOST_IS_WINDOWS", isWindows
             buildConfigField "int", "VERSION_NUM", "0"
+            buildConfigField "int", "APP_ID", localProperties.get("TELEGRAM_APP_ID", TELEGRAM_APP_ID)
+            buildConfigField "String", "APP_HASH", "\"" + localProperties.get("TELEGRAM_APP_HASH", TELEGRAM_APP_HASH) + "\""
         }
 
         HA_private {
@@ -139,12 +148,14 @@ android {
             proguardFiles getDefaultProguardFile('proguard-android.txt'), '../TMessagesProj/proguard-rules.pro', '../TMessagesProj/proguard-rules-beta.pro'
             ndk.debugSymbolLevel = 'FULL'
             buildConfigField "String", "BUILD_VERSION_STRING", "\"" + APP_VERSION_NAME + "\""
-            buildConfigField "String", "APP_CENTER_HASH", "\"" + getProps("APP_CENTER_HASH_PRIVATE") + "\""
-            buildConfigField "String", "BETA_URL", "\"" + getProps("BETA_PRIVATE_URL") + "\""
+            buildConfigField "String", "APP_CENTER_HASH", "\"" + localProperties.get("APP_CENTER_HASH_PRIVATE", "") + "\""
+            buildConfigField "String", "BETA_URL", "\"" + localProperties.get("BETA_PRIVATE_URL", "") + "\""
             buildConfigField "boolean", "DEBUG_VERSION", "true"
             buildConfigField "boolean", "DEBUG_PRIVATE_VERSION", "true"
             buildConfigField "boolean", "BUILD_HOST_IS_WINDOWS", isWindows
             buildConfigField "int", "VERSION_NUM", "1"
+            buildConfigField "int", "APP_ID", localProperties.get("TELEGRAM_APP_ID", TELEGRAM_APP_ID)
+            buildConfigField "String", "APP_HASH", "\"" + localProperties.get("TELEGRAM_APP_HASH", TELEGRAM_APP_HASH) + "\""
         }
 
         HA_public {
@@ -154,12 +165,14 @@ android {
             proguardFiles getDefaultProguardFile('proguard-android.txt'), '../TMessagesProj/proguard-rules.pro', '../TMessagesProj/proguard-rules-beta.pro'
             ndk.debugSymbolLevel = 'FULL'
             buildConfigField "String", "BUILD_VERSION_STRING", "\"" + APP_VERSION_NAME + "\""
-            buildConfigField "String", "APP_CENTER_HASH", "\"" + getProps("APP_CENTER_HASH_PUBLIC") + "\""
-            buildConfigField "String", "BETA_URL", "\"" + getProps("BETA_PUBLIC_URL") + "\""
+            buildConfigField "String", "APP_CENTER_HASH", "\"" + localProperties.get("APP_CENTER_HASH_PUBLIC", "") + "\""
+            buildConfigField "String", "BETA_URL", "\"" + localProperties.get("BETA_PUBLIC_URL", "") + "\""
             buildConfigField "boolean", "DEBUG_VERSION", "true"
             buildConfigField "boolean", "DEBUG_PRIVATE_VERSION", "false"
             buildConfigField "boolean", "BUILD_HOST_IS_WINDOWS", isWindows
             buildConfigField "int", "VERSION_NUM", "4"
+            buildConfigField "int", "APP_ID", localProperties.get("TELEGRAM_APP_ID", TELEGRAM_APP_ID)
+            buildConfigField "String", "APP_HASH", "\"" + localProperties.get("TELEGRAM_APP_HASH", TELEGRAM_APP_HASH) + "\""
         }
 
         HA_hardcore {
@@ -169,12 +182,14 @@ android {
             proguardFiles getDefaultProguardFile('proguard-android.txt'), '../TMessagesProj/proguard-rules.pro', '../TMessagesProj/proguard-rules-beta.pro'
             ndk.debugSymbolLevel = 'FULL'
             buildConfigField "String", "BUILD_VERSION_STRING", "\"" + APP_VERSION_NAME + "\""
-            buildConfigField "String", "APP_CENTER_HASH", "\"" + getProps("APP_CENTER_HASH_HARDCORE") + "\""
-            buildConfigField "String", "BETA_URL", "\"" + getProps("BETA_HARDCORE_URL") + "\""
+            buildConfigField "String", "APP_CENTER_HASH", "\"" + localProperties.get("APP_CENTER_HASH_HARDCORE", "") + "\""
+            buildConfigField "String", "BETA_URL", "\"" + localProperties.get("BETA_HARDCORE_URL", "") + "\""
             buildConfigField "boolean", "DEBUG_VERSION", "true"
             buildConfigField "boolean", "DEBUG_PRIVATE_VERSION", "true"
             buildConfigField "boolean", "BUILD_HOST_IS_WINDOWS", isWindows
             buildConfigField "int", "VERSION_NUM", "5"
+            buildConfigField "int", "APP_ID", localProperties.get("TELEGRAM_APP_ID", TELEGRAM_APP_ID)
+            buildConfigField "String", "APP_HASH", "\"" + localProperties.get("TELEGRAM_APP_HASH", TELEGRAM_APP_HASH) + "\""
         }
 
         standalone {
@@ -190,6 +205,8 @@ android {
             buildConfigField "boolean", "DEBUG_PRIVATE_VERSION", "false"
             buildConfigField "boolean", "BUILD_HOST_IS_WINDOWS", isWindows
             buildConfigField "int", "VERSION_NUM", "6"
+            buildConfigField "int", "APP_ID", localProperties.get("TELEGRAM_APP_ID", TELEGRAM_APP_ID)
+            buildConfigField "String", "APP_HASH", "\"" + localProperties.get("TELEGRAM_APP_HASH", TELEGRAM_APP_HASH) + "\""
         }
 
         release {
@@ -206,23 +223,14 @@ android {
             buildConfigField "boolean", "DEBUG_PRIVATE_VERSION", "false"
             buildConfigField "boolean", "BUILD_HOST_IS_WINDOWS", isWindows
             buildConfigField "int", "VERSION_NUM", "7"
+            buildConfigField "int", "APP_ID", localProperties.get("TELEGRAM_APP_ID", TELEGRAM_APP_ID)
+            buildConfigField "String", "APP_HASH", "\"" + localProperties.get("TELEGRAM_APP_HASH", TELEGRAM_APP_HASH) + "\""
         }
     }
 
     namespace 'org.telegram.messenger'
 }
 
-
-def getProps(String propName) {
-    def propsFile = rootProject.file('local.properties')
-    if (propsFile.exists()) {
-        def props = new Properties()
-        props.load(new FileInputStream(propsFile))
-        return props[propName]
-    } else {
-        return "";
-    }
-}
 apply plugin: 'com.google.gms.google-services'
 
 task checkVisibility {
diff --git a/TMessagesProj/src/main/java/org/telegram/messenger/BuildVars.java b/TMessagesProj/src/main/java/org/telegram/messenger/BuildVars.java
index 4ed7e6491..a3034fe1b 100644
--- a/TMessagesProj/src/main/java/org/telegram/messenger/BuildVars.java
+++ b/TMessagesProj/src/main/java/org/telegram/messenger/BuildVars.java
@@ -26,8 +26,8 @@ public class BuildVars {
     public static boolean NO_SCOPED_STORAGE = Build.VERSION.SDK_INT <= 29;
     public static String BUILD_VERSION_STRING = BuildConfig.BUILD_VERSION_STRING;
 
-    public static int APP_ID = 4;
-    public static String APP_HASH = "014b35b6184100b085b0d0572f9b5103";
+    public static int APP_ID = BuildConfig.APP_ID;
+    public static String APP_HASH = BuildConfig.APP_HASH;
 
     // SafetyNet key for Google Identity SDK, set it to empty to disable
     public static String SAFETYNET_KEY = "AIzaSyDqt8P-7F7CPCseMkOiVRgb1LY8RN1bvH8";
diff --git a/TMessagesProj_App/build.gradle b/TMessagesProj_App/build.gradle
index 5cb2449e5..a42f4ca0a 100644
--- a/TMessagesProj_App/build.gradle
+++ b/TMessagesProj_App/build.gradle
@@ -42,18 +42,25 @@ android {
     }
 
     signingConfigs {
+        def localProperties = new Properties()
+        def localPropertiesFile = rootProject.file("local.properties")
+
+        if (localPropertiesFile.exists()) {
+            localProperties.load(new FileInputStream(localPropertiesFile))
+        }
+
         debug {
-            storeFile file("../TMessagesProj/config/release.keystore")
-            storePassword RELEASE_STORE_PASSWORD
-            keyAlias RELEASE_KEY_ALIAS
-            keyPassword RELEASE_KEY_PASSWORD
+            storeFile rootProject.file(localProperties.get("RELEASE_STORE_FILE", RELEASE_STORE_FILE))
+            storePassword localProperties.get("RELEASE_STORE_PASSWORD", RELEASE_STORE_PASSWORD)
+            keyAlias localProperties.get("RELEASE_KEY_ALIAS", RELEASE_KEY_ALIAS)
+            keyPassword localProperties.get("RELEASE_KEY_PASSWORD", RELEASE_KEY_PASSWORD)
         }
 
         release {
-            storeFile file("../TMessagesProj/config/release.keystore")
-            storePassword RELEASE_STORE_PASSWORD
-            keyAlias RELEASE_KEY_ALIAS
-            keyPassword RELEASE_KEY_PASSWORD
+            storeFile rootProject.file(localProperties.get("RELEASE_STORE_FILE", RELEASE_STORE_FILE))
+            storePassword localProperties.get("RELEASE_STORE_PASSWORD", RELEASE_STORE_PASSWORD)
+            keyAlias localProperties.get("RELEASE_KEY_ALIAS", RELEASE_KEY_ALIAS)
+            keyPassword localProperties.get("RELEASE_KEY_PASSWORD", RELEASE_KEY_PASSWORD)
         }
     }
 
diff --git a/TMessagesProj_AppHockeyApp/build.gradle b/TMessagesProj_AppHockeyApp/build.gradle
index fb9af6a75..85281e204 100644
--- a/TMessagesProj_AppHockeyApp/build.gradle
+++ b/TMessagesProj_AppHockeyApp/build.gradle
@@ -55,18 +55,25 @@ android {
     }
 
     signingConfigs {
+        def localProperties = new Properties()
+        def localPropertiesFile = rootProject.file("local.properties")
+
+        if (localPropertiesFile.exists()) {
+            localProperties.load(new FileInputStream(localPropertiesFile))
+        }
+
         debug {
-            storeFile file("../TMessagesProj/config/release.keystore")
-            storePassword RELEASE_STORE_PASSWORD
-            keyAlias RELEASE_KEY_ALIAS
-            keyPassword RELEASE_KEY_PASSWORD
+            storeFile rootProject.file(localProperties.get("RELEASE_STORE_FILE", RELEASE_STORE_FILE))
+            storePassword localProperties.get("RELEASE_STORE_PASSWORD", RELEASE_STORE_PASSWORD)
+            keyAlias localProperties.get("RELEASE_KEY_ALIAS", RELEASE_KEY_ALIAS)
+            keyPassword localProperties.get("RELEASE_KEY_PASSWORD", RELEASE_KEY_PASSWORD)
         }
 
         release {
-            storeFile file("../TMessagesProj/config/release.keystore")
-            storePassword RELEASE_STORE_PASSWORD
-            keyAlias RELEASE_KEY_ALIAS
-            keyPassword RELEASE_KEY_PASSWORD
+            storeFile rootProject.file(localProperties.get("RELEASE_STORE_FILE", RELEASE_STORE_FILE))
+            storePassword localProperties.get("RELEASE_STORE_PASSWORD", RELEASE_STORE_PASSWORD)
+            keyAlias localProperties.get("RELEASE_KEY_ALIAS", RELEASE_KEY_ALIAS)
+            keyPassword localProperties.get("RELEASE_KEY_PASSWORD", RELEASE_KEY_PASSWORD)
         }
     }
 
diff --git a/TMessagesProj_AppHuawei/build.gradle b/TMessagesProj_AppHuawei/build.gradle
index 8c77aa960..149fa3f9c 100644
--- a/TMessagesProj_AppHuawei/build.gradle
+++ b/TMessagesProj_AppHuawei/build.gradle
@@ -50,18 +50,25 @@ android {
     }
 
     signingConfigs {
+        def localProperties = new Properties()
+        def localPropertiesFile = rootProject.file("local.properties")
+
+        if (localPropertiesFile.exists()) {
+            localProperties.load(new FileInputStream(localPropertiesFile))
+        }
+
         debug {
-            storeFile file("../TMessagesProj/config/release.keystore")
-            storePassword RELEASE_STORE_PASSWORD
-            keyAlias RELEASE_KEY_ALIAS
-            keyPassword RELEASE_KEY_PASSWORD
+            storeFile rootProject.file(localProperties.get("RELEASE_STORE_FILE", RELEASE_STORE_FILE))
+            storePassword localProperties.get("RELEASE_STORE_PASSWORD", RELEASE_STORE_PASSWORD)
+            keyAlias localProperties.get("RELEASE_KEY_ALIAS", RELEASE_KEY_ALIAS)
+            keyPassword localProperties.get("RELEASE_KEY_PASSWORD", RELEASE_KEY_PASSWORD)
         }
 
         release {
-            storeFile file("../TMessagesProj/config/release.keystore")
-            storePassword RELEASE_STORE_PASSWORD
-            keyAlias RELEASE_KEY_ALIAS
-            keyPassword RELEASE_KEY_PASSWORD
+            storeFile rootProject.file(localProperties.get("RELEASE_STORE_FILE", RELEASE_STORE_FILE))
+            storePassword localProperties.get("RELEASE_STORE_PASSWORD", RELEASE_STORE_PASSWORD)
+            keyAlias localProperties.get("RELEASE_KEY_ALIAS", RELEASE_KEY_ALIAS)
+            keyPassword localProperties.get("RELEASE_KEY_PASSWORD", RELEASE_KEY_PASSWORD)
         }
     }
 
diff --git a/TMessagesProj_AppStandalone/build.gradle b/TMessagesProj_AppStandalone/build.gradle
index b06efd6ac..904e368da 100644
--- a/TMessagesProj_AppStandalone/build.gradle
+++ b/TMessagesProj_AppStandalone/build.gradle
@@ -44,18 +44,25 @@ android {
     }
 
     signingConfigs {
+        def localProperties = new Properties()
+        def localPropertiesFile = rootProject.file("local.properties")
+
+        if (localPropertiesFile.exists()) {
+            localProperties.load(new FileInputStream(localPropertiesFile))
+        }
+
         debug {
-            storeFile file("../TMessagesProj/config/release.keystore")
-            storePassword RELEASE_STORE_PASSWORD
-            keyAlias RELEASE_KEY_ALIAS
-            keyPassword RELEASE_KEY_PASSWORD
+            storeFile rootProject.file(localProperties.get("RELEASE_STORE_FILE", RELEASE_STORE_FILE))
+            storePassword localProperties.get("RELEASE_STORE_PASSWORD", RELEASE_STORE_PASSWORD)
+            keyAlias localProperties.get("RELEASE_KEY_ALIAS", RELEASE_KEY_ALIAS)
+            keyPassword localProperties.get("RELEASE_KEY_PASSWORD", RELEASE_KEY_PASSWORD)
         }
 
         release {
-            storeFile file("../TMessagesProj/config/release.keystore")
-            storePassword RELEASE_STORE_PASSWORD
-            keyAlias RELEASE_KEY_ALIAS
-            keyPassword RELEASE_KEY_PASSWORD
+            storeFile rootProject.file(localProperties.get("RELEASE_STORE_FILE", RELEASE_STORE_FILE))
+            storePassword localProperties.get("RELEASE_STORE_PASSWORD", RELEASE_STORE_PASSWORD)
+            keyAlias localProperties.get("RELEASE_KEY_ALIAS", RELEASE_KEY_ALIAS)
+            keyPassword localProperties.get("RELEASE_KEY_PASSWORD", RELEASE_KEY_PASSWORD)
         }
     }
 
diff --git a/TMessagesProj_AppTests/build.gradle b/TMessagesProj_AppTests/build.gradle
index 33480a996..9bd4c2203 100644
--- a/TMessagesProj_AppTests/build.gradle
+++ b/TMessagesProj_AppTests/build.gradle
@@ -54,18 +54,25 @@ android {
     }
 
     signingConfigs {
+        def localProperties = new Properties()
+        def localPropertiesFile = rootProject.file("local.properties")
+
+        if (localPropertiesFile.exists()) {
+            localProperties.load(new FileInputStream(localPropertiesFile))
+        }
+
         debug {
-            storeFile file("../TMessagesProj/config/release.keystore")
-            storePassword RELEASE_STORE_PASSWORD
-            keyAlias RELEASE_KEY_ALIAS
-            keyPassword RELEASE_KEY_PASSWORD
+            storeFile rootProject.file(localProperties.get("RELEASE_STORE_FILE", RELEASE_STORE_FILE))
+            storePassword localProperties.get("RELEASE_STORE_PASSWORD", RELEASE_STORE_PASSWORD)
+            keyAlias localProperties.get("RELEASE_KEY_ALIAS", RELEASE_KEY_ALIAS)
+            keyPassword localProperties.get("RELEASE_KEY_PASSWORD", RELEASE_KEY_PASSWORD)
         }
 
         release {
-            storeFile file("../TMessagesProj/config/release.keystore")
-            storePassword RELEASE_STORE_PASSWORD
-            keyAlias RELEASE_KEY_ALIAS
-            keyPassword RELEASE_KEY_PASSWORD
+            storeFile rootProject.file(localProperties.get("RELEASE_STORE_FILE", RELEASE_STORE_FILE))
+            storePassword localProperties.get("RELEASE_STORE_PASSWORD", RELEASE_STORE_PASSWORD)
+            keyAlias localProperties.get("RELEASE_KEY_ALIAS", RELEASE_KEY_ALIAS)
+            keyPassword localProperties.get("RELEASE_KEY_PASSWORD", RELEASE_KEY_PASSWORD)
         }
     }
 
diff --git a/gradle.properties b/gradle.properties
index 6bb315db1..d4aefb394 100644
--- a/gradle.properties
+++ b/gradle.properties
@@ -19,7 +19,10 @@ APP_PACKAGE=org.telegram.messenger
 IS_PRIVATE=false
 RELEASE_KEY_PASSWORD=android
 RELEASE_KEY_ALIAS=androidkey
+RELEASE_STORE_FILE=TMessagesProj/config/release.keystore
 RELEASE_STORE_PASSWORD=android
+TELEGRAM_APP_ID=4
+TELEGRAM_APP_HASH=014b35b6184100b085b0d0572f9b5103
 org.gradle.jvmargs=-Xmx4096M
 org.gradle.daemon=true
 org.gradle.parallel=true
